### ISSUE TITLE
supporting recursion

### ISSUE THREAD

**Issue by afilinski** [Ginko-X/Streaming_NESL#20]
- should dynamic graph extension happen in withctrl or scall?
  - should scall be completely equivalent to unfolding (with graph
      only elaborated by need, possibly in withctrl)?

**Comment by afilinski** [comment]
notes from meeting:

syntax:
  WithCtrl SId [Sid] [Sid] [Instr]  -- new ctrl stream, imports, exports (maybe STree for now), code
  
-- snesl source
```
  function f(x:int):int = x*(x+2)

  reducePlus({-f(x+5) : x in &3}) : int -- or replace the 3 with a 0
```
-- svcode target 
```
  f :=
    SFun [S1] [S4]
      [S0 := Ctrl -- needed?
       S2 := Const 2
       S3 := MapTwo Add S1 S2
       S4 := MapTwo Mul S1 S3]

  it :=
    SFun [] [S11]
      [S0 := Ctrl -- <*>
       S1 := Const 3 -- <3>
       S2 := ToFlags S1  -- <F,F,F,T>
       S3 := Usum S2 -- <*,*,*>
       S4 := MapConst S3 1 -- <1,1,1>
       S5 := SegscanPlus S4 S2 <0,1,2>
       S6 := Usum S2 -- <*,*,*>
       WithCtrl S6 [S5] [S10]
         [S7 := Const 5 -- <5,5,5>
          S8 := MapTwo Add S5 S7 -- <5,6,7>
          Call F [S8] [S9]  -- S9 := <35,48,63>
          S10 := Map Negate S9] -- <-35,-48,-63>
       S11 := ReducePlus S10 S2] -- <-146>
```
multiple users of s: `let s = &3 in sum({f(x): x in s}) * sum(s)`


initial dag:
  Instead of withctrl block make
    * pseudo-client of S5 which will never input anything,
      but blocked on new Ctrl stream?
    * pseudo-producer of S10 which will never output anything (not even eos),
      but blocked on new Ctrl stream?

once S6 is non-empty (or EOS flag), replace pseudo-clients

  if S6 was empty then

    * remove pseudo-client of S5
    * replace pseudo-producer of S10 with (global?) producer of empty stream

  else

    * generate the graph for the WithCtrl body (unfolding the call to F!) and
    * replace pseudo-clients and pseudo-producers with actual clients & producers

**Comment by Ginko-X** [comment]
The latest update at this point can somehow support some "simple" recursions. The code is still messy and unreadable. 

Here are some of my questions.

1. 
In #16 , you mentioned something about the new SId system of runtime, `RSId`, when adding recursion:
> Another, weakly related issue: maybe you shouldn't use the same type SId for both "source code" stream ids in SExp and "runtime" stream ids in Svctx, because when you add recursion, they will definitely no longer be the same. Maybe for now you can just have a type alias RSId = SId, but so that in the future it will be easy to replace RSId with something else, such as RSId = (Int, SId), corresponding to a unique stack-frame number and an offset within that frame, like in the usual compilation scheme for sequential languages with recursion...

What I think I should have in my code is 
```Haskell
type RSId = (Int,SId,SId)
```
The first `Int` is stack-frame number and the last `SId` the offset as you suggested. But in some cases these two may be not enough, e.g., when we try to call the same function multiple times: `f(1) + 2 + f(3)` if we consider `f(1)` and `f(2)` are at the same calling-depth level. So I add the middle one, which is the head of `retSIds` in `SCall f argSIds retSIds`, because the return SIds must be unique for function calls at the same level. Another reason is it doesn't require a sequential interpreting order of the streams.

2. A runnable simple case of recursion can be like:
```Haskell
function f(x:int):int = if x >= 5 then x else f(x+1)

> {f(x) : x in &20}
{5,5,5,5,5,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19} :: {int}
```
But it easily crashed for more  general examples (buffersize = 10):
- fibonacci numbers, premature Eos exception when `x` > 3, can be a bug in code
```Haskell
function fibo(x:int) :int = if x < 2 then 1 else fibo(x-1)+fibo(x-2)
```

- `uc_scan` deadlocks when #v >=16, maybe a true deadlock?
```Haskell
-- unite-and-conquer (exclusive) scan
-- only for a sequence with length of power of 2
function uc_scan(v:{int}) : {int} =
  if empty(v) then {}int else 
  let is = scanExPlus({1 : x in v})
    in if reducePlus(is) == 0 then {0} else
         let odds = concat({{x|i%2==0} :  i in is, x in v});
             evens = concat({{x|not(i%2 == 0)} : i in is, x in v});
             ps = {x+y : x in odds, y in evens}
          in concat({{s,s+x} : s in uc_scan(ps), x in odds})
```

**Comment by afilinski** [comment]
from the meeting:
```
 function f(x:int):int = if x==0 then 1 else x*f(x-1)

> :bs 100000
Buffer size: 100000
> {f(x): x in &10}
{1,1,2,6,24,120,720,5040,40320,362880} :: {int}
[W_H: 763, S_H: 121], [W_L: 12603039, S_L: 335]
W_L/W_H: 16517.7, S_L/S_H: 2.8
S_L/(W_H/bufSize + S_H): 2.8
> :bs 1
Buffer size: 1
> {f(x): x in &10}
{1,1,2,6,24,120,720,5040,40320,362880} :: {int}
[W_H: 763, S_H: 121], [W_L: 3084, S_L: 1144]
W_L/W_H: 4.0, S_L/S_H: 9.5
S_L/(W_H/bufSize + S_H): 1.3
```

**Comment by afilinski** [comment]
remember to make a streamable version of empty(), i.e., not consume the entire stream before saying T or F.

**Comment by afilinski** [comment]
I think the deadlock in the scan() that we talked about yesterday is actually because it was coded in such a way that the first element of the output (which is always 0) can only be produced after we have consumed the first element of the input stream, because the conditional tests whether the stream of odd-numbered elements is non-empty, so as a minimum, we need to get element number 0 out of the way first. 

There are a couple of ways around that, but they all get a little ugly, especially when we don't have many stream-processing primitives in the language. But here's a simple solution, which assumes that we know the length of the stream from the start, so that all the conditionals in the recursion can be resolved without looking at the actual stream elements at all. I have also extended it to compute both the (exclusive) scan and the reduction at the same time:
```function scanred(v:{int},n:int) : ({int},int) =  
   let is = scanExPlus({1 : x in v});
       evens = concat({{x|i%2==0} :  i in is, x in v});
       odds = concat({{x|not(i%2 == 0)} : i in is, x in v})
   in if n==1 then ({0}, reducePlus(v)) -- reducePlus() acts like the() here
      else let ps = {x+y : x in evens, y in odds};	
               (ss,r) = scanred(ps,n/2)
           in (concat({{s,s+x} : s in ss, x in evens}), r)
```
This seems to work, e.g., with `scanred(&16, 16)`, even for `:bs 1`. 
Note that the `reducePlus(v)` should really have been `the(v)` (they return the same when the length of the stream is exactly 1), but if I put `the(v)` in the code, it still deadlocks on long streams, which seems to be a bug unrelated to recursion, but possibly with the handling of conditionals. In particular, this also deadlocks:
```let s = &100 in if F then s else s```
but, strangely, this does not:
```let s = &100 in if T then s else s```

**Comment by afilinski** [comment]
Argh, I accidentally omitted the first line from the code in my previous comment. It should read:
```
function scanred(v:{int}, n:int) : ({int},int) =  
   let is = scanExPlus({1 : x in v});
       evens = concat({{x|i%2==0} :  i in is, x in v});
       odds = concat({{x|not(i%2 == 0)} : i in is, x in v})
   in if n==1 then ({0}, the(v))
      else let ps = {x+y : x in evens, y in odds};	
               (ss,r) = scanred(ps,n/2)
           in (concat({{s,s+x} : s in ss, x in evens}), r)
```
And with the latest fixes, it seems to now work correctly also with `the(v)` in the n==1 case. Great!

**Comment by afilinski** [comment]
Incidentally, consider adding a little more syntactic sugar to the parser:

* `e1 != e2` should desugar to `not(e1 == e2)` (much like you already do for `e1 < e2`)
* `{e : x1 in e1, ..., xn in en | e0}` should desugar to `concat({{e|e0} : x1 in e1, ..., xn in en}`

This will make the scan examples a little nicer to read and write.

**Comment by Ginko-X** [comment]
By now I have fixed all the known bugs about recursion.

- About scan 
> I think the deadlock in the scan() that we talked about yesterday is actually because it was coded in such a way that the first element of the output (which is always 0) can only be produced after we have consumed the first element of the input stream ...

I think this reason is more convincing. 
I tried to code inclusive scan as you said (on this Tuesday) it would not deadlock as the exclusive one does. But it also deadlocks, if I code in the following way:
```Haskell
function in_scan(v:{int}) : {int} =
    let is = scanExPlus({1 : x in v});
        odds = concat({{x|not(i%2==0)} :  i in is, x in v});
        evens = concat({{x|i%2 == 0} : i in is, x in v})
    in if empty(odds) then evens else
       let ps = {x+y : x in odds, y in evens}
       in concat({{s-x,s} : s in in_scan(ps), x in odds})
```
And yes, your solution `scanred` works. Can we still say this algorithm is streamable? Or is there a definition of **streamable**?

 - About RSId 
My earlier version `type RSId = (Int,SId,SId)` in the 2nd comment is not correct. For the instruction `SCall fid argSids retSids`, `retSids` are unique for identifying each call in the current stack frame, and therefore used as the 2nd component of `RSId` (actually just `head retSids`). But it is still possible to have some RSId collisions in deeper stack frames if the recursive function calls itself multi-times inside itself, such as the fibonacci example mentioned before. So after applying a new calculation method of the 2nd component, the bug in the fibonacci code is fixed.


- Interpreter performance issue in the 3rd comment (the cost bug has been fixed)
I think this is mainly caused by stealing. When bs = 1, no stealing at all. But when bs > 1,  stealing can happen frequently especially when the data is irregular. And the stealing algorithm is inefficient: when a deadlock happens, it tries to find the first buffer-non-empty processor and switch it to draining mode. In most cases, switching just one processor is not enough; this procedure keeps looping until the deadlock is broken or throws a real-deadlock exception, so the number of scheduling round increases (maybe) linearly to the length of `Svctx` (runtime context). In recursion, the situation becomes much worse as the length of Svctx also increases dynamically. Take the fibonacci code as an example, the command `:r` (in the interactive interpreter) shows that `fibo(4)` uses 3 rounds with bs = 1 VS 328 rounds with bs = 10.

So I think the interpreter performance could be improved after we implement a better scheduling strategy, such as the two-phase scheduling.  Should we just give it a lower priority now?

- In case you're interested, here is the reason for the bug `let s = &100 in if F then s else s` deadlocks while `let s = &100 in if T then s else s` does not :
For `WithCtrl ctrl ims code retSIds`, when `ctrl` is detected empty,  the `retSIds` processors are shut down immediately. What I ignored is that they should also be deleted from their suppliers' client list if have, otherwise it will cause a deadlock. As in most cases the suppliers of these `retSIds` are also inside `code`, this special case where `retSIds` are defined even before WithCtrl is what I never considered. 

- Other questions:
    - I also tried to implement a streaming quicksort, but I didn't find a way to do that. Is it not streamable at all? If not, is there any streamable recursive sorting algorithm?
    - By now, I'm not quite sure if there is still some big bug in the code about recursion. (Also, the performance and scheduling are not good enough.) But if we give adding vectors a higher priority (i.e., #22), I think I should just start to look at that then?

**Comment by afilinski** [comment]
Some comments:

* Re: scans. Your `in_scan` code actually suffers from a similar problem as the original `uc_scan`. To be streamable, it should produce each element of the output stream immediately after consuming the corresponding element of the input stream `v`. But in `{s-x,s}`, where `s in in_scan(ps)`, the first element of `ps` (and therefore also of `scan(ps)`) actually depends on the first _two_ elements of `v`, and hence we cannot output `s-x` when we are supposed to.
Actually, using subtraction in a plus-scan is a bit of a cheat in any case, because it won't generalize to other associative operators, such as max, or even multiplication (because multiplication by 0 is not invertible). It should be possible to express an inclusive scan using the combining operator only (it doesn't even need to have a neutral element), but I'm not sure if we have enough stream-handling primitives in the current language to do so.

* Re: streamability. I think a reasonable definition of a streamable algorithm is precisely one that can be executed in constant space (or, more generally, in space linear in the recursion depth). But it may be that the current set of stream-processing primitives is not sufficient to express all streamable algorithms as streamable SNESL programs. So by that definition, scanred() is streamable, and it's certainly an improvement on plain NESL, because we only need to know the length of the sequence up front, but we don't need to have it materialized all at once.
Whether the variant that determines the length automatically from the sequence itself can also be expressed in the current SNESL, I'm not sure, especially in the general case where the length is not necessarily a power of two. (But if we are given the length of the sequence explicitly, it's trivial to pad it at the end to a power of 2, use the special-case algorithm, and then discard the extra padding; and both the padding and the trimming can clearly also be done in a streaming fashion.)

* Re: RSId. OK, that makes sense.

* Re: performance. I don't think stealing is the whole explanation. In particular, when all buffers are sufficiently large, no stealing should be necessary at all. Yet the fibonacci example runs significantly slower (both according to work/step costs and wall-clock time) with any bs>1 than bs=1. Also, with the way we have defined work cost, shouldn't it be completely independent of the buffer size? Buffer sizes should only affect the step cost, because sometimes we can do more of the work in a single step.  And even if a lot of stealing happens, it may increase the scheduling overhead (and hence wall-clock runtime), but it should not require a larger total number of steps than for bs=1. I think there's still some non-trivial work to be done here, even if we don't care so much about actual performance, only the cost model.

* Re: spurious deadlocks. OK, makes sense. The code for handling WithControl looks really complicated right now, but it does seem to handle all situations.

* Re: quicksort. I don't think any sorting algorithm can really be streaming, since (almost by definition) it needs to consume the entire input before it can start producing the output, just in case that the last input element is actually the smallest. So I think we'll need to have (at least non-nested) vectors working to do quicksort.

* Re: vectors. After thinking about it some more, nested vectors may actually not be so problematic after all. But I think that the most important right now is getting the recursion performance and cost-model anomalies sorted out (at least partially), and possibly even get started on the proofs, rather than adding even more features. The fact that scanred() can actually be made streaming shows that it's possible to do non-trivial examples involving recursion and sequences, so it's not so critical to also have vectors in the language (though I still hope we'll have time to add them).

* Re: syntactic sugar. Very good!

**Comment by Ginko-X** [comment]
OK. Then I'll investigate more about the left issues here about recursion,  rather than start #22 .

**Comment by Ginko-X** [comment]
- Interpreter performance
After thinking more about the execution of the fibonacci code, I found that the main reason why it became even slower when bs > 1 is the last chunk "stealing": as most of the streams' xducers only produce one chunk, which is just size 1 and less than the buffer size, switching those buffers caused more scheduling rounds. So I removed the`Eos` state and modified the type of `BufState` as we've discussed before:
```Haskell
data BufState = Filling [AVal] 
              | Draining [AVal] Bool   
```
Now the fibonacci code performs more reasonable: when bs >1, only 1 scheduling round is needed, so it runs faster.

There should be still some other reasons that affect the performance, such as stealing, for which I don't see a minor modification of the code is enough.

- Cost model 
> with the way we have defined work cost, shouldn't it be completely independent of the buffer size?

I also fixed the cost model a little, so now, yes work cost is independent of bs; bs only affects the step cost.

But so far I also see a strange case, `scanred`, where as `bs` increases, work also increases a little bit. I haven't figured out the reason.
```
> :bs 1
Buffer size: 1
>
> scanred(&32,32)
({0,0,1,3,6,10,15,21,28,36,45,55,66,78,91,105,120,136,153,171,190,210,231,253,276,300,325,351,378,406,435,465},496) :: ({int},int)
[W_H: 1246, S_H: 190], [W_L: 7837, S_L: 3129]
W_L/W_H: 6.3
S_L/(W_H/bufSize + S_H): 2.2

> :bs 10
> scanred(&32,32)
[W_H: 1246, S_H: 190], [W_L: 7851, S_L: 778]  -- work increases 14
W_L/W_H: 6.3
S_L/(W_H/bufSize + S_H): 2.5
>
> :bs 20
> scanred(&32,32)
[W_H: 1246, S_H: 190], [W_L: 7871, S_L: 491]  -- work increases 20
W_L/W_H: 6.3
S_L/(W_H/bufSize + S_H): 1.9
>
> :bs 100000
> scanred(&32,32)
[W_H: 1246, S_H: 190], [W_L: 7955, S_L: 383] 
W_L/W_H: 6.4
S_L/(W_H/bufSize + S_H): 2.0
```

Besides, it may be worth noting that this cost model for the streaming interpreter (let's call it Streaming model) also has several differences from the one for the eager interpreter (Eager model):

1. The Eager model uses a function `wrapWork` (i.e., counting work from 1, not 0) on its work to avoid the case where work < step (also to keep consistent with the SNESL high-level cost model); but in Streaming model this case will not happen, because when both the input and output are empty (therefore 0 work), step is also 0 (because no buffer state switch). So it's not necessary to use `wrapWork` in Streaming model.

Another special case is the SExp `Check`, which only reads from inputs but no outputs, so the step in Eager mode _S_e_ is 1, while the step in Streaming model _S_s_ is 0  

2. Function calls have extra cost in Streaming model. For simplicity, when passing parameters, I used a special xducer `rinOut` which just passes the input stream and does nothing on it, to connect the streams of the formal and actual parameters: 
```Haskell
rinOut = rin 0 (\x -> rout x >> rinOut)
```
This method is also used on the pass of ctrl and return streams. As a result, it brings the extra cost of both work and step. 
I think this usually will only add constant cost, or maybe linear in the worst case, but we may discuss this tomorrow.

So combining these two points above,  when `bs` is infinitely large and **no function calls** in SVCODE, W_s (work of Streaming model) is usually less than W_e (work of Eager model) by a const _n_, where _n_ is the number of SVCODE instructions (or more precisely, the number of stream definitions), and S_s ≤ S_e. 
But when **there are some function calls** in SVCODE (and `bs` is still infinite), which model costs more will depend on the concrete SVCODE program, but the difference should be still relatively small.

I'm not sure if it's meaningful to compare these Streaming model and Eager model in such detail, because our goal, I think, is mainly to keep only the Streaming one consistent with the SNESL high level one.

**Comment by afilinski** [comment]
The work-cost anomaly in scanred() seems to have something to do with the interpreter sometimes finishing too early, when there is still (useless) work left to do:
```
> :bs 1
Buffer size: 1
> let v=&100 in {v|F}
{} :: {{int}}
[W_H: 104, S_H: 4], [W_L: 1018, S_L: 505]
W_L/W_H: 9.8
S_L/(W_H/bufSize + S_H): 4.7
> :bs 2
Buffer size: 2
> let v=&100 in {v|F}
{} :: {{int}}
[W_H: 104, S_H: 4], [W_L: 1021, S_L: 255]
W_L/W_H: 9.8
S_L/(W_H/bufSize + S_H): 4.6
> :bs 10000
Buffer size: 10000
> let v=&100 in {v|F}
{} :: {{int}}
[W_H: 104, S_H: 4], [W_L: 1316, S_L: 9]
W_L/W_H: 12.7
S_L/(W_H/bufSize + S_H): 2.2
```
Note that if we make the comprehension body more complicated, the work with bs=1 goes up to the value for bs=infininity, but the step cost (correctly) remains unaffected:
```
> :bs 1
Buffer size: 1
> let v=&100 in {v++v|F}
{} :: {{int}}
[W_H: 104, S_H: 4], [W_L: 1316, S_L: 505]
W_L/W_H: 12.7
S_L/(W_H/bufSize + S_H): 4.7
```
Maybe that'll help you narrow down the problem.

**Comment by Ginko-X** [comment]
I think this cost bug is still about the situation where return SIds are defined before the WithCtrl block. How much (useless) work these streams will do depends on the time when they are shut down. 
I'll try to find a solution later. Thank you, this is helpful.

**Comment by Ginko-X** [comment]
Should we also put the streams of `Distr`, `Pack` and `UPack` generated in comprehensions into WithCtrl block (by modifying the compiler code)?
For instance, if we  modify the SVCODE for exp `let v = &100 in {reducePlus(v) |F}` from Figure 1 to Figure 2, it seems make more sense that when we know the new ctrl is empty, it's not necessary to do the distribution or pack tasks at all, (and we only do that when the new ctrl is non-empty)?
```
-- Figure 1
S0 := Ctrl;
S1 := Const 100;
S2 := ToFlags S1;
S3 := Usum S2;
S4 := MapConst S3 1;
S5 := SegscanPlus S4 S2;
S6 := Const F;
S7 := B2u S6;
S8 := Usum S7;  -- new ctrl

S9 := Distr S6 S2;  
S10 := Pack S5 S9;
S11 := UPack S2 S6;
WithCtrl S8 (import []):
        S12 := ReducePlus S10 S11
        Return: (IStr S12)
Return: (SStr (IStr S12), S7)
```

```
-- Figure 2
S0 := Ctrl;
...     
S7 := B2u S6;
S8 := Usum S7;

WithCtrl S8 (import []):  -- not necessary to change `import`
        S9 := Distr S6 S2;  -- move them inside withctrl
        S10 := Pack S5 S9;
        S11 := UPack S2 S6;

	S12 := ReducePlus S10 S11
	Return: (IStr S12)
Return: (SStr (IStr S12), S7)
```

-------- updated  on Aug 9

Because in the exp `let v = &100 in {v |F}` , it is a kind of paradox that, on one hand, we must compute the full streams of Pack or UPack (i.e. S9 and S10), on the other hand, according to the semantics of WithCtrl, we must shut down them immediately (therefore no cost on them since then)  when we know the new ctrl is empty as they are the return SIds of this WithCtrl block. 
```
S0 := Ctrl;
...
S8 := Distr S6 S2;
S9 := Pack S5 S8;
S10 := UPack S2 S6;
S11 := Usum S7;    -- it doesn't matter whether the new ctrl is before or after Distr, Pack and UPack
WithCtrl S11 (import []):
	Return: (SStr (IStr S9), S10)  -- return SIds
Return: (SStr (SStr (IStr S9), S10), S7)

**Comment by afilinski** [comment]
First, I think there may be a problem with the import list of the WithCtrl. With the currently checked-in compiler, I get:
```
> :c let v=&100 in {v|F}

Parameters: []
S0 := Ctrl; 
S1 := Const 100; 
S2 := ToFlags 1; 
S3 := Usum 2; 
S4 := MapConst 3 1; 
S5 := SegscanPlus 4 2; 
S6 := Const F; 
S7 := B2u 6; 
S8 := Distr 6 2; 
S9 := Pack 5 8; 
S10 := UPack 2 6; 
S11 := Usum 7; 
WithCtrl S11 (import []):
	Return: (SStr (IStr 9), 10)
Return: (SStr (SStr (IStr 9), 10), 7)
```
That is, the import list is empty even though S9 and S10 are meant to be used in the WithCtrl code (even if only to be exported again). On the other hand, if I replace `{v|F}` with `{v++v|F}` or `{reducePlus(v)|F}`, the import list is computed correctly. I can't tell if this is the cause of the problem with premature shutdown, but I think that in general, the import list should be based on what the WithCtrl body can potentially use, rather than what it actually uses, because the latter is impossible to compute accurately anyway without unfolding function calls at compile time, and you end up getting subtly different behavior in cases like this:
```
> function id(x:{int}):{int} = x
Defined function: id
> :c let v=&100 in {v|F}
...
WithCtrl S11 (import []):
...
> :c let v=&100 in {id(v)|F}
...
WithCtrl S11 (import [9,10]):  
...
```
I think you should just compute the import list based on newVarTrs, rather than by using getImportSids.

Now as for whether it would make sense to move the Pack into the WithCtrl, I think that wouldn't really gain much efficiency in practice, because if we are doing more than one restricted comprehension in parallel, we only hit the empty WithCtrl case if _all_ of them have a false guard, which is not going to be all that common. More importantly, the high-level cost model says that we have to pay for the Pack in all cases anyway, so it's probably not worth it to modify the compilation scheme to do a little better in some cases. Finally, it's currently the case that if WithCtrl is invoked with an empty control stream, then all of the streams on the import list are also guaranteed to be empty, which would no longer be the case if we moved the pack into the body of WithCtrl. I'm not sure if that property is actually exploited, but it's possible that moving the Pack into WithCtrl, so that it can only start consuming once we know that tne new control stream is  definitely non-empty, might cause a deadlock. My advice would be to leave it alone for now.

**Comment by afilinski** [comment]
Oh, and something unrelated that I just thought of: When generating the flowgraph from compiled vcode, and especially when unfolding Scalls, you might think of this as a further (runtime) compilation-like process, and in particular, you can naturally organize it around a graph-generation environment mapping SIds to RSIds, just like the compiler itself maps Snesl variable names to (trees of) SIds. In particular, then you can let RSIds be just integers, and include any information about which SIds they originally corresponded to as part of the graph node instead, since that information isn't actually used for anything other than visualization and debugging. Making RSIds as simple to compare as possible should also help speed up lookups and updates, whether you use Data.Map or just a plan association list.

**Comment by Ginko-X** [comment]
> More importantly, the high-level cost model says that we have to pay for the Pack in all cases anyway, ...

I think I made a mistake here. What I have in the current high-level cost model for `{e_0 | e_1}` is:
 - if `e_1` ↓ T, cost (W(e_1) + W(e_0) +1, S(e_1) + S(e_0) +1)
 - otherwise `e_1` ↓ F,  cost (W(e_1) +1 , S(e_1) +1) 

So I didn't consider the Pack  (and UPack, Distr) cost in sequence comprehensions at all in the high-level cost model.

My concern in that question is if we move Pack inside WithCtrl, the (streaming interpreter) code could be simplified more or less, because in that way the return SIds would only be defined inside WithCtrl. (But this will also involve a fix of the high-level cost model, and may cause a deadlock as you pointed out? I'm not sure.)

About the import list. I also noticed that problem before; the reason why I didn't put those return SIds into the import list is that I think it's not necessary to make a process a pseudo-producer at the same time also add it a pseudo-client (since no xducer can produce anything if it doesn't consume anything, except for `Ctrl` and `EmptyCtrl`). 

Anyway, for now, I'll just try to fix this cost bug without any other big change.

**Comment by Ginko-X** [comment]
> Oh, and something unrelated that I just thought of: When generating the flowgraph from compiled vcode, and especially when unfolding Scalls, you might think of this as a further (runtime) compilation-like process, and in particular, you can naturally organize it around a graph-generation environment mapping SIds to RSIds, just like the compiler itself maps Snesl variable names to (trees of) SIds. In particular, then you can let RSIds be just integers... 

This paragraph doesn't make sense. By "let RSIds be just integers" do you mean `type RSId = Int` instead of a pair or triple of `Int`? If so, then it sounds like that you're suggesting to dynamically generate RSIds (using fresh SIds)?

**Comment by afilinski** [comment]
Yes, you can take RSId = Int (or maybe `newtype RSid = R Int deriving (Eq, Show)`, to make sure that you never confuse SId and RSId), and dynamically allocate fresh RSId's when building the graph from Svcode, just like you dynamically create SId's during compilation of Snesl to Svcode. And in particular, this graph construction will be analogous to the version of the compiler that dynamically unfolds function calls in Snesl code to generate just linear Svcode, so that no traces of the Scall remain in the generated graph.

On the other hand, when you generate code for a WithCtrl instruction, you'll need to create a special process node that becomes a client of the new control stream, and that keeps a copy of the SId->RSid graph-building environment, as well as the svcode from the WithCtrl body. Then, once that special process gets activated during execution (i.e., when the control stream becomes available for draining), you can resume construction of the graph exactly where you left off, and patch it in instead of the pseudo-clients and pseudo-producers.

If this is still unclear, you're welcome to stop by my office today (before 17 or so).

**Comment by Ginko-X** [comment]
I have a better understanding of your idea after yesterday's discussion now.
But how about moving on to start some part of the formal proof if there is no immediate crash? I guess knowing more details about what/how I should prove would give me a better idea of how I should fix the code. 
What do you think?

**Comment by afilinski** [comment]
I guess you may as well start working on the proofs (for the language without user-defined functions), in parallel with fixing up the code. The first step in proving anything about cost preservation, of course, will be to formalize _on paper_ the semantics (including time and space costs) of the source and target languages. The source semantics is more or less done (except for the couple of bugs we talked about), but we don't really have anything like a formal semantics of the target language, which will probably need to be expressed in small-step style, to properly account for the concurrency aspects. Maybe you can start thinking about that for our meeting on Tuesday? Perhaps assume for now that buffers are unlimited, so that we don't have to worry about deadlocks, whether apparent (i.e., resolvable by stealing) or real.

However, there is at least one serious bug having to do specifically with functions in the current interpreter:
```
> :bs 10
Buffer size: 10
> {}int
{} :: {int}
[W_H: 1, S_H: 1], [W_L: 3, S_L: 2]
W_L/W_H: 3.0
S_L/(W_H/bufSize + S_H): 1.8
> function id(x:{int}):{int}=x
Defined function: id
> id({}int)
Deadlock!
> id(&0)
{} :: {int}
[W_H: 2, S_H: 2], [W_L: 13, S_L: 6]
W_L/W_H: 6.5
S_L/(W_H/bufSize + S_H): 2.7
```
You should probably fix that first.

**Comment by Ginko-X** [comment]
> However, there is at least one serious bug having to do specifically with functions in the current interpreter: 
....

Bug fixed.

**Comment by Ginko-X** [comment]
When I reviewed the code, I got some idea to remove the unnecessary Xduers for parameter passing during unfolding SCalls, which ended up with something quite similar to what you told me last Friday, that is, adding a runtime environment mapping from (R)SIds to RSIds, (but for now I'm still using triples for RSIds).

The code (although still kind of messy) looks simpler for SCall instruction init and interp, and hopefully we could remove it from `sInstrInterp` when it completely works, (then we may see how to fix `WithCtrl`.) Also, now no unexpected extra cost any more.

But it's still buggy, at least doesn't work for the function `scanred`.

----- updated on Aug, 15
It works for `scanred` now.

**Comment by Ginko-X** [comment]
> On the other hand, when you generate code for a WithCtrl instruction, you'll need to create a special process node that becomes a client of the new control stream and that keeps a copy of the SId->RSid graph-building environment, as well as the svcode from the WithCtrl body. Then, once that special process gets activated during execution (i.e., when the control stream becomes available for draining), you can resume construction of the graph exactly where you left off, and patch it in instead of the pseudo-clients and pseudo-producers.

I have some similar idea but does not require to set up a special process (or modify the type of `Process`). Instead, I need to add a new component 
```Haskell
type UEnv = [(RSId, (Bool, [RSId], [RSId], [SIntsr]))]  -- this Bool is like an unfolding-flag
```
to the runtime monad, then: 
  
- In the Init phase for `WithCtrl ctrl ins outs code`: 
    -  add `(ctrl, (F, ins, tree2sids outs,code))` to the UEnv
    - other usual initialization including pseudo-clients and pseudo-producers setting-up

- In the Interp phase: 
    - interp a `rsid` as usual
    - check if this `rsid` is a key in UEnv, 
         - no || yes && unfolding-flag is `T`, do nothing 
         - yes && flag is `F`, detect if it's empty or not, and then decide whether to unfold the code or not, and update the flag with `T` when necessary

This method is based on an assumption: each WithCtrl instruction has a unique ctrl RSId. 
As a ctrl RSId is always newly generated when we construct a WithCtrl instruction looking from the compiler code,  I think this assumption holds, so this method should be feasible.  

I'm wondering what's your opinion. Please let me know if any.

**Comment by afilinski** [comment]
(I'll have to think a bit more about this one. More tomorrow.)

**Comment by afilinski** [comment]
OK, I'm still not sure that I understand the proposed setup entirely, but I think that it could work. However, I don't know if it would also fit properly with a two-phase scheduler, especially in the queueing-based model, where it is the responsibility of the process filling a buffer to actively notify all blocked clients that the buffer is ready for draining (and conversely, the last client to drain a buffer must notify the producer that the buffer is again ready for filling). The setup I had in mind was primarily meant for that setting, i.e., the process actually filling the control-stream buffer (or marking it as Eos), typically USum, would wake up the special process associated with the WithCtrl, and that process would do the relevant on-demand graph generation and pseudo-client/supplier replacement, without the scheduler or other parts of the interpreter even being aware of it, so as to keep all the complexity of supporting WithCtrl in one place.

In any case, though, I think that as long as your current scheduler does not contain any known bugs (unexpected deadlocks and the like), it should be safe to leave it alone for a while, while you are working on the correctness proofs - at least the value and work-cost parts. When we get to step-costs and space usage, we may need to take a closer look at the whole scheduling area again.

I think that the way to get the most benefit out of all the coding you have done would actually be to do some experiments with a more heavily instrumented Svcode interpreter. That is, extend it to collect statistics on how many scheduling rounds were needed, how often you had to steal a buffer, what the average size of a buffer was when it was stolen, whether the choice of scheduling strategy matters, and the like - and then summarize and discuss those results in the thesis.  That would quite possibly be more valuable than yet another correctness result for step-costs (which I'm not even sure holds, if it turns out that we do a lot of stealing), and would also add a real experimental dimension to your thesis - but without all the practical complications of measuring actual wall-clock runtimes, where you also have to account for caching/paging effects, hardware/compiler peculiarities, interference from other tasks on the machine, and the like.

**Comment by Ginko-X** [comment]
OK, these all sound reasonable. Thank you.
